src/Build/Graph/ProjectGraph.cs

```diff
@@ -742,16 +745,31 @@ private static ImmutableList<string> ExpandDefaultTargets(ImmutableList<string>
             return targets;
         }
 
-        internal static ProjectInstance DefaultProjectInstanceFactory(
+        internal ProjectInstance DefaultProjectInstanceFactory(
```

Comment:
Might be overthinking this, but this seems the `EvaluationContext` in scope for the lifetime of the graph when it's only needed for graph construction.

Consider something like this instead, which is effectively what we do for QuickBuild:

```
if (projectInstanceFactory == null)
{
    EvaluationContext evaluationContext = EvaluationContext.Create(EvaluationContext.SharingPolicy.Shared);
   projectInstanceFactory = (string projectPath, Dictionary<string, string> globalProperties, ProjectCollection projectCollection) =>new ProjectInstance(
                projectPath,
                globalProperties,
                MSBuildConstants.CurrentToolsVersion,
                subToolsetVersion: null,
                projectCollection,
                evaluationContext);
}
```

Reply:
I guess the `ProjectInstance` objects hold onto the `EvaluationContext` anyway, so probably doesn't matter. I guess it avoids one `EvaluationContext` allocation if a factory isn't provided, but that's pretty small potatoes.

Reply:
This would keep the existing allocations for the existing cases where the factory is provided, I like it.

Reply:
Oh actually there's a smaller code change to accomplish the same goal, which I'm going to take: only initialize the context if necessary.